// genkitx-openai/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var src_exports = {};
__export(src_exports, {
  dallE3: () => import_dalle.dallE3,
  default: () => src_default,
  gpt35Turbo: () => import_gpt.gpt35Turbo,
  gpt4: () => import_gpt.gpt4,
  gpt41: () => import_gpt.gpt41,
  gpt41Mini: () => import_gpt.gpt41Mini,
  gpt41Nano: () => import_gpt.gpt41Nano,
  gpt45: () => import_gpt.gpt45,
  gpt4Turbo: () => import_gpt.gpt4Turbo,
  gpt4Vision: () => import_gpt.gpt4Vision,
  gpt4o: () => import_gpt.gpt4o,
  gpt4oMini: () => import_gpt.gpt4oMini,
  gpt4oMiniTts: () => import_tts.gpt4oMiniTts,
  gpt4oTranscribe: () => import_whisper.gpt4oTranscribe,
  o1: () => import_gpt.o1,
  o1Mini: () => import_gpt.o1Mini,
  o1Preview: () => import_gpt.o1Preview,
  o3: () => import_gpt.o3,
  o3Mini: () => import_gpt.o3Mini,
  o4Mini: () => import_gpt.o4Mini,
  openAI: () => openAI,
  textEmbedding3Large: () => import_embedder.textEmbedding3Large,
  textEmbedding3Small: () => import_embedder.textEmbedding3Small,
  textEmbeddingAda002: () => import_embedder.textEmbeddingAda002,
  tts1: () => import_tts.tts1,
  tts1Hd: () => import_tts.tts1Hd,
  whisper1: () => import_whisper.whisper1
});
module.exports = __toCommonJS(src_exports);
var import_plugin = require("genkit/plugin");
var import_openai = require("openai");
var import_dalle = require("./dalle.js");
var import_embedder = require("./embedder.js");
var import_gpt = require("./gpt.js");
var import_tts = require("./tts.js");
var import_whisper = require("./whisper.js");
const openAI = (options) => (0, import_plugin.genkitPlugin)("openai", (ai) => __async(void 0, null, function* () {
  var _a;
  const client = new import_openai.OpenAI(options);
  for (const name of Object.keys(import_gpt.SUPPORTED_GPT_MODELS)) {
    (0, import_gpt.gptModel)(ai, name, client);
  }
  (_a = options == null ? void 0 : options.models) == null ? void 0 : _a.map((model) => {
    if (!model.name || !model.info || !model.configSchema) {
      throw new Error(`Model ${model.name} is missing required fields`);
    }
    (0, import_gpt.gptModel)(ai, model.name, client, model.info, model.configSchema);
  });
  (0, import_dalle.dallE3Model)(ai, client);
  for (const name of Object.keys(import_whisper.SUPPORTED_STT_MODELS)) {
    (0, import_whisper.sttModel)(ai, name, client);
  }
  for (const name of Object.keys(import_tts.SUPPORTED_TTS_MODELS)) {
    (0, import_tts.ttsModel)(ai, name, client);
  }
  for (const name of Object.keys(import_embedder.SUPPORTED_EMBEDDING_MODELS)) {
    (0, import_embedder.openaiEmbedder)(ai, name, options);
  }
}));
var src_default = openAI;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  dallE3,
  gpt35Turbo,
  gpt4,
  gpt41,
  gpt41Mini,
  gpt41Nano,
  gpt45,
  gpt4Turbo,
  gpt4Vision,
  gpt4o,
  gpt4oMini,
  gpt4oMiniTts,
  gpt4oTranscribe,
  o1,
  o1Mini,
  o1Preview,
  o3,
  o3Mini,
  o4Mini,
  openAI,
  textEmbedding3Large,
  textEmbedding3Small,
  textEmbeddingAda002,
  tts1,
  tts1Hd,
  whisper1
});
